### PR TITLE
[ML][Inference] compressing model definition and lazy parsing

### PR DESCRIPTION
**Pull request by benwtrent**
This PR adds the initial stab of compressed model format storage. 

The model definition is always stored as a compressed string in the index. When the user's request the trained model config, by default it will supply the inflated object. If they request with `decompress_definition=false` then the compressed string is returned instead. 

There is still the overall issue of how we handle arbitrarily large models when users want to export/import them. But this is a step in the right direction.

NOTE: It does not add the logic to split between documents.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        return Streams.readFully(inflateStream);
+    }
+
+    //Public for testing (for now)
+    public static String deflate(BytesReference reference) throws IOException {
+        BytesStreamOutput out = new BytesStreamOutput();
+        try (OutputStream compressedOutput = new GZIPOutputStream(out, BUFFER_SIZE)) {
+            reference.writeTo(compressedOutput);
+        }
+        return new String(Base64.getEncoder().encode(BytesReference.toBytes(out.bytes())), StandardCharsets.UTF_8);
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java]: I would add a header/magic, e.g. just 1 byte. This will ensure future proof if you at some point in future want to change the format, which can be different compression, different baseXX encoding or something else. If you think base64 will never be changed the header can be in out and e.g. start with `01`, otherwise you need a printable/json-friendly version. 

The other thing is a truncation guard. I think neither gzipoutputstream nor base64 is safe regarding truncation. So the header should store the length, I would use a variable length encoding for that.

In summary:

{COMPRESSION_HEADER_BYTE}{VINT_ENCODED_LENGTH}{COMPRESSED_DEFINITION}

inflate of course needs the counterpart, compression header is just a check for now, in future this might branch to different implementations. The encoded length however should be checked, so you do not even have to try to parse it if you find a mismatch.

Additionally a the length should have a safe guard, with other words: define a maximum allowed length.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfigTests.java [hunk 2]
```diff
@@ -110,27 +119,69 @@ protected boolean assertToXContentEquivalence() {
...
+        reference = XContentHelper.toXContent(config,
+            XContentType.JSON,
+            new ToXContent.MapParams(Collections.singletonMap("human", "false")),
+            false);
+        assertThat(reference.utf8ToString(), not(containsString("\"definition\"")));
+        assertThat(reference.utf8ToString(), containsString("compressed_definition"));
+        assertThat(reference.utf8ToString(), containsString(lazyModelDefinition.getCompressedString()));
+    }
+
+    public void testParseWithBothDefinitionAndCompressedSupplied() throws IOException {
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfigTests.java]: :+1: 

:imp: 

testParseWithBothDefinitionAndCompressedSuppliedIncompatible()
testParseCompressedSuppliedTruncated()
testParseCompressedSuppliedGarbage()

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ * Collection of helper methods. Similar to CompressedXContent, but this utilizes GZIP for parity with the native compression
+ */
+public final class ToXContentCompressor {
+    private static final int BUFFER_SIZE = 4096;
+
+    private ToXContentCompressor() {}
+
+    public static <T extends ToXContentObject> String deflate(T objectToCompress) throws IOException {
+        BytesReference reference = XContentHelper.toXContent(objectToCompress, XContentType.JSON, false);
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java]: why JSON? A binary encoding might be better suited, we support CBOR and SMILE ootb. It might not make a difference in terms of size but in terms of parsing it back (runtime).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            LoggingDeprecationHandler.INSTANCE,
+            inflate(compressedString),
+            XContentType.JSON)) {
+            return parser.mapOrdered();
+        }
+    }
+
+    static BytesReference inflate(String compressedString) throws IOException {
+        byte[] compressedBytes = Base64.getDecoder().decode(compressedString.getBytes(StandardCharsets.UTF_8));
+        InputStream inflateStream = new GZIPInputStream(new BytesArray(compressedBytes).streamInput(), BUFFER_SIZE);
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java]: I think a `BoundedInputStream` (apache commons) should be used to protect for zip bombs. The generic ToMap parser can be abused to cause a OOM and crash the node.

**Review by hendrikmuhs — COMMENTED** [review]
added some comments and concerns

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfig.java [hunk 5]
```diff
@@ -73,6 +78,17 @@
...
+            (p, c) ->
+            {
+                if (p.currentToken() == XContentParser.Token.START_OBJECT) {
+                    return LazyModelDefinition.fromParsedDefinition(TrainedModelDefinition.fromXContent(p, ignoreUnknownFields).build());
+                }
+                throw new IllegalArgumentException(
+                    "unexpected token [" + p.currentToken() + "] for [" + DEFINITION.getPreferredName() + "]");
+            },
+            DEFINITION);
+        parser.declareString(TrainedModelConfig.Builder::setLazyDefinition, COMPRESSED_DEFINITION);
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfig.java]: Supplying both causes it to throw. I don't think we should ever allow both to be set. 

As for when we expect one over the other, I am not sure. 

I do think optionally allowing the compressed definition to be returned is beneficial to the user as it gives a better stop-gap for moving models between clusters while we work on an import/export feature.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/FrequencyEncoding.java [hunk 6]
```diff
@@ -66,7 +67,7 @@ public static FrequencyEncoding fromXContentLenient(XContentParser parser) {
     public FrequencyEncoding(String field, String featureName, Map<String, Double> frequencyMap) {
         this.field = ExceptionsHelper.requireNonNull(field, FIELD);
         this.featureName = ExceptionsHelper.requireNonNull(featureName, FEATURE_NAME);
-        this.frequencyMap = Collections.unmodifiableMap(ExceptionsHelper.requireNonNull(frequencyMap, FREQUENCY_MAP));
+        this.frequencyMap = Collections.unmodifiableMap(new LinkedHashMap<>(ExceptionsHelper.requireNonNull(frequencyMap, FREQUENCY_MAP)));
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/FrequencyEncoding.java]: I thought I removed this from the PR (it was leftover troubleshooting). Please ignore :). Will make sure the change is removed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Map;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Collection of helper methods. Similar to CompressedXContent, but this utilizes GZIP for parity with the native compression
+ */
+public final class ToXContentCompressor {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java]: I thought keeping it in the inference package was good enough. But I can rename it.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java (hunk 3 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java]: The reason for JSON was future interop with native. I think it would be nice to have a definition that is trivially supported by the native process so we can simply pass the compressed strings without any processing.

**Comment by benwtrent** [comment]
run elasticsearch-ci/packaging-sample-matrix

**Review by hendrikmuhs — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java (hunk 3 above)
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java]: :+1:  I forgot that this needs native support, at least SMILE has no good implementation for C++. LBNL parsing should not be the bottleneck.

**Comment by benwtrent** [comment]
run elasticsearch-ci/2
